internal/controllers/machine/machine_controller.go

```diff
@@ -655,16 +746,10 @@ func (r *Reconciler) drainNode(ctx context.Context, cluster *clusterv1.Cluster,
 		GracePeriodSeconds:  -1,
 		// If a pod is not evicted in 20 seconds, retry the eviction next time the
 		// machine gets reconciled again (to allow other machines to be reconciled).
-		Timeout: 20 * time.Second,
-		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
-			verbStr := "Deleted"
-			if usingEviction {
-				verbStr = "Evicted"
-			}
-			log.Info(fmt.Sprintf("%s pod from Node", verbStr),
-				"Pod", klog.KObj(pod))
-		},
-		Out: writer{log.Info},
+		Timeout:                         20 * time.Second,
```

Comment:
[blocker]

Okay I now spent considerable time reading through the drainHelper and I fail to see how we could use these two funcs to write a reliable status condition.

They are explicitly documented to be used for printing progress output:
> // OnPodDeletionOrEvictionFinished is called when a pod is eviction/deletetion is failed; for printing progress output
not to calculate conditions. And I think I know why

RunNodeDrain roughly works like this:
* Call GetPodsForDeletion
* Call DeleteOrEvictPods
  * Call evictPods with the list of pods (in our case always this func, never deletePods)
    * evictPods starts a goroutine for *every* single Pod and then goes through:
      * trigger eviction
      * wait for Pod deletion

The result being that across all those goroutines the OnPodDeletionOrEvictionStarted and OnPodDeletionOrEvictionFinished funcs are called in almost random order.

Noteworthy detail: there are even cases where for a specific Pod only OnPodDeletionOrEvictionStarted will be called and then never OnPodDeletionOrEvictionFinished

Based on this I think our condition will have pretty random content.


Am I missing something?

Reply:
Just to give an example of a proper API that we could build on top:

It would be great if the RunNodeDrain func just has a response parameter that provides us exactly the info we need, which e.g. could be a list of Pods & eviction status that we can iterate through and see if eviction was started / succeeded / failed or not.

Reply:
To provide more context. We have to implement this in a way that the condition stays 100% the same if nothing on the Node changes. Otherwise we will continuously update the condition and probably end up with a reconcile ever 10-20ms (Machine status updates trigger another reconcile)

Reply:
When OnPodDeletionOrEvictionFinished returns error can we just show that in the status? Other callbacks I agree we can keep in the logs.
The issue that we are trying to resolve here is that when some pod gets stuck in draining user can see it in the machine conditions so user can see it easily, finding that information in logs is difficult.

Reply:
Is there any guarantee that OnPodDeletionOrEvictionFinished is only called once per drain? (and that if we repeat the drain and nothing happens in the workload error we end up with the same error again?)

Reply:
when draining is stuck for example because of pdb we will get an error from kube-drain every time it tries to drain, and it will keep on trying until manual intervention. But we can check on CAPI side if the failure condition is set and we get an error from kube-drain again we don't set it again.

Reply:
Which means the information on the condition gets outdated? (let's assume we try to evict a few pods and at some point another eviction works, we wouldn't update the condition then becuse it is already set?)

Reply:
When we get a successful callback of the same pod that was stuck we can remove the condition. I am proposing to set conditions only when the pod is stuck., and remove the condition when that pod is fixed(e,g pdb is removed by manual intervention) and eventually evicted. It is just a way to inform the user that this pod is stuck and needs some kind of manual intervention for draining to succeed.

Reply:
I have made some changes:

- No longer setting any condition when OnPodDeletionOrEvictionStarted, only printing logs.
- The condition is only set when OnPodDeletionOrEvictionFinished fail/pass.

Reply:
@sbueringer please take a look.

Reply:
I think the current version still does not address the following problems:
1. The OnPodDeletionOrEvictionFinished func is documented to be only used for printing logs
    ```go
	// OnPodDeletionOrEvictionFinished is called when a pod is eviction/deletetion is failed; for printing progress output
	OnPodDeletionOrEvictionFinished func(pod *corev1.Pod, usingEviction bool, err error)
   ```
   It is clearly not an API to get reliable deterministic behavior from, please also see: https://github.com/kubernetes-sigs/cluster-api/pull/10104#discussion_r1584629274
2. The current code leads to flickering conditions. Example:
    * There are 5 Pods pod-1, .. pod-5 that cannot be evicted due to PDBs
    * 1st Reconcile => condition message will be "Failed to evict pod pod-1 from node node-1"
    * 2nd Reconcile => condition message will be "Failed to evict pod pod-4 from node node-1" (there is no guarantee in which order onPodDeletionOrEvictionFinished will be called for all the Pods that are being evicted)
    * => This means that the conditions will be continuously updated with varying messages => this leads to **a lot** of reconciles for all controllers watching the Node object
    * => The message only surfaces a subset of the Pods that we're not able to evict. 
3. Let's make another example:
   * We have 5 pods again (pod-1, ... pod-5) , on one reconcile we are able to only evict pod-3 all the other evictions fail (i.e. drainErr != nil)
   * It's possible that the condition message will now be "Evicted pod pod-3 from node node-1"
   * I think that's highly misleading because the reason why `DrainingSucceeded` is false is not that we could evict pod-3, but that we couldn't evict the other 4 pods

To be honest, I simply think with the current implementation of the drainHelper there is no way to surface a good condition message. It simply does not provide the API for it.

Reply:
@sbueringer 

> The current code leads to flickering conditions

I really would prefer flickering conditions to not having any Condition.

If a node fails to drain, then I expect to see a Condition and an Event.

Is there anything I can do to help implement this?

Reply:
I'm going to work on this pretty soon.

Some folks might prefer flickering conditions that lead to permanent reconciles of all controllers watching Machine objects, but I don't. And I think we can't merge something that leads to effects like this

Reply:
> I'm going to work on this pretty soon.

thank you very much!

Just one note: I don't like flickering conditions, too. But in this case the node gets drained. This means that the resource will be deleted soon. As soon as the reason which blocks the drain is removed, the resource is gone and the condition, too.

Please tell me if, I can support you somehow!

Reply:
Sure, definitely will ping you on the PR to take a long and maybe try it out :)

The problem is that there is no guarantee that a drain finishes.

Let's assume:
* a nodeDrainTimeout is not set (the default)
* PDBs are used in the workload cluster
* evictions that are done as part of drain are denied because of PDBs
* the application that uses the PDBs doesn't recover itself / come up on new nodes

In that situation a drain can be stack for a long time (even if it's just a few hours or a few days it's not ideal).

If in that timestamp the situation doesn't change on the Node (i.e. the exact some Pods are still on the Node but cannot be further drained) we have to write a deterministic condition. Otherwise the Machine controller will just permanently reconcile (Machine controller writes conditions => update event => Machine reconcile is triggered => repeat)

Reply:
@sbueringer thank you very much for explaining that.

About this endless circle: Machine controller writes conditions => update event => Machine reconcile is triggered 

Why not filter via `WithEventFilter` and a predicate?

Reply:
That would help for this controller, yup

Reply:
(But this is not my only concern about this implementation, see my previous comments above. That's why I prefer a new clean implementation)